### ISSUE TITLE
SSL And HTTP Requests show as HTTP/1.1

### ISSUE THREAD

**Issue by webhost247** [engintron/engintron#437]
Hello, After installing engintron on my cPanel server Domains installed under my server are reporting network requests within Chrome and Firefox as HTTP/1.1 and not HTTP/2

Could someone please explain the cause of this, I thought engintron served requests on domains using HTTP/2 now

**Comment by Nirjonadda** [comment]
Yes, This same issue we have. @fevangelou  Engintron are not Enabled http2 support?

**Comment by tux-tn** [comment]
As i said for @Nirjonadda  issue, it's related to Openssl version shipped in CentOS, you can find more details and a solution here: #368

**Comment by Nirjonadda** [comment]
@tux-tn I does found the solution that you are posted. Please send me centos 7 installation guide for HTTP/2

**Comment by ubidzz** [comment]
The problem is the OpenSSL version that is installed in your server because http/2 needs OpenSSL 1.0.2 and CentOS and cPanel installs OpenSSL/1.0.1e-fips. And it not a good idea to upgrade your OpenSSL because it will brake things in your server so the only way to get OpenSSL 1.0.2 installed it will have to be a standalone OpenSSL just for the nginx server so that your server is running 2 different OpenSSL versions. Then setting up the nginx server to use the newer OpenSSL and not 1.0.1e. 

I don't even know if you can run 2 different OpenSSL on a server with out any problems.

**Comment by webhost247** [comment]
@ubidzz @tux-tn My worry is that replacing cPanels OpenSSL will cause a lot of problems.. @tux-tn can you comment on how everything in cPanel works after updating OpenSSL?

**Comment by ubidzz** [comment]
i did what @tux-tn for comment #https://github.com/engintron/engintron/issues/368 and seems to work and it didn't mess with my cPanel openSSL. Now all my SSL websites use http/2.

**Comment by webhost247** [comment]
After following comment #368 (Installing the repo) I then followed up by doing "yum makecache" and then tried doing "yum update" but no packages get marked for updating.. aka OpenSSL

**Comment by ubidzz** [comment]
I ran the command they gave me on the site to add the repo to the server then went in to WHM -> Software -> System Update with Include kernel packages and it will update using the new repo then remove and install Engintron then restart your server.

**Comment by tux-tn** [comment]
@webhost247  the package i mentioned is not replacing OpenSSL version from CentOS repos.
Actually it's a custom nginx build with an embed custom and more recent OpenSSL version. 
If you add the repo, it will only update nginx and won't break anything else.
Here is an example from one server where i enabled the CodeIT repo:
```shell
[root@nevermind ~]# openssl version
OpenSSL 1.0.1e-fips 11 Feb 2013
```
```shell
[root@lucy ~]# nginx -V
nginx version: nginx/1.11.10
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-11) (GCC) 
built with OpenSSL 1.0.2k  26 Jan 2017
TLS SNI support enabled
```
Note the difference between the version reported by OpenSSL and the one reported by nginx.
So to resume things: enabling the repo will only update nginx and add support to ALPN/H2

**Comment by tux-tn** [comment]
@webhost247  the package i mentioned is not replacing OpenSSL version from CentOS repos.
Actually it's a custom nginx build with an embed custom and more recent OpenSSL version. 
If you add the repo, it will only update nginx and won't break anything else.
Here is an example from one server where i enabled the CodeIT repo:
```shell
[root@nevermind ~]# openssl version
OpenSSL 1.0.1e-fips 11 Feb 2013
```
```shell
[root@nevermind ~]# nginx -V
nginx version: nginx/1.11.10
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-11) (GCC) 
built with OpenSSL 1.0.2k  26 Jan 2017
TLS SNI support enabled
```
Note the difference between the version reported by OpenSSL and the one reported by nginx.
So to resume things: enabling the repo will only update nginx and add support to ALPN/H2

**Comment by ubidzz** [comment]
@tux-tn do you also use their Apache HTTPd?

**Comment by tux-tn** [comment]
@ubidzz No, i do not see the need if nginx will serve all the users requests

**Comment by Nirjonadda** [comment]
@tux-tn If this steps are correct?

Download and install the public signing key:

rpm --import https://repo.codeit.guru/RPM-GPG-KEY-codeit

Added the following in my /etc/yum.repos.d/ directory in a file with a .repo suffix, for example codeit.repo

[CodeITmainline]
name=CodeIT mainline repo
baseurl=https://repo.codeit.guru/packages/mainline/centos/7/$basearch
enabled=1
gpgkey=https://repo.codeit.guru/RPM-GPG-KEY-codeit
gpgcheck=1

yum install OpenSSL

or

yum update

**Comment by ubidzz** [comment]
no
1. SSH in your server
2. cd /etc/yum.repos.d && wget https://repo.codeit.guru/codeit.el`rpm -q --qf "%{VERSION}" $(rpm -q --whatprovides redhat-release)`.repo
3. Go to WHM -> Software -> System Update with Include kernel packages
4  Restart your server
5.  Reinstall Engintron
6. Restart the nginx server

**Comment by webhost247** [comment]
@ubidzz After adding the repo I went to WHM and tried doing a System Update as advised but it says

![image](https://cloud.githubusercontent.com/assets/5007268/23335086/182d36be-fba5-11e6-8b30-b49c9bcbbfb4.png)

**Comment by tux-tn** [comment]
You don't need to reinstall Engintron

Run the command provided by codeIT to add the correct repo and update the system/install the new nginx version:

For nginx mainline:
```shell
cd /etc/yum.repos.d && wget https://repo.codeit.guru/codeit.mainline.el`rpm -q --qf "%{VERSION}" $(rpm -q --whatprovides redhat-release)`.repo && yum update
```

For nginx stable:
```shell
cd /etc/yum.repos.d && wget https://repo.codeit.guru/codeit.el`rpm -q --qf "%{VERSION}" $(rpm -q --whatprovides redhat-release)`.repo && yum update
```

**Comment by Nirjonadda** [comment]
@tux-tn is this OpenSSL For nginx stable?

Updated:
  apr.x86_64 0:1.5.2-1.el7.codeit

Complete!

**Comment by tux-tn** [comment]
@Nirjonadda Yes, it's the latest stable OpenSSL release

can you provide the output of `yum info nginx` to see if the version from CodeIT has been installed or not?

**Comment by Nirjonadda** [comment]
@tux-tn Just installed but still can not use engintron for this issue. https://github.com/engintron/engintron/issues/439

This issue happened with option "Edit your custom_rules for Nginx" Not working nginx with any custom rules. What happening?

This rules for Auto logged out of Xenforo.

> if ($SITE_URI ~* "mysite.com|mysite.com/login|mysite.com/register|mysite.com/logout|mysite.com/admin.php") {
     set $CACHE_BYPASS_FOR_DYNAMIC 1; # Disables micro-caching
  # set $CACHE_BYPASS_FOR_STATIC 1; # Disables static file caching

This rules for non-www URL.

> if ($host = 'mydomain.com' ){
    rewrite ^/(.*)$ $scheme://www.mydomain.com/$1 permanent;
}

**Comment by Nirjonadda** [comment]
@tux-tn Still are using HTTP/1.1 , Please let me know the fix.

``
[root@na ~]# yum info nginx
Loaded plugins: fastestmirror, tsflags, universal-hooks
Loading mirror speeds from cached hostfile
 * EA4: 185.69.232.245
 * EA4-experimental: 185.69.232.245
 * base: mirrors.coreix.net
 * epel: mirror.math.princeton.edu
 * extras: mirrors.coreix.net
 * updates: mirrors.coreix.net
Installed Packages
Name        : nginx
Arch        : x86_64
Epoch       : 1
Version     : 1.10.3
Release     : 1.el7.ngx
Size        : 2.3 M
Repo        : installed
From repo   : nginx
Summary     : High performance web server
URL         : http://nginx.org/
License     : 2-clause BSD-like license
Description : nginx [engine x] is an HTTP and reverse proxy server, as well as
            : a mail proxy server.

[root@na ~]#
``

**Comment by webhost247** [comment]
I have tried running as advised  "cd /etc/yum.repos.d && wget https://repo.codeit.guru/codeit.mainline.el`rpm -q --qf "%{VERSION}" $(rpm -q --whatprovides redhat-release)`.repo && yum update"

My server does not install anything new after it finishes adding the Repo, I have also tried going into cPanel and running "System Update" but nothing seams to be working. 

My output of nginx -V is :

> [root@server yum.repos.d]# nginx -V
nginx version: nginx/1.11.10
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC)
built with OpenSSL 1.0.1e-fips 11 Feb 2013[root@server

**Comment by Nirjonadda** [comment]
@webhost247 Please do run yum clean all then run yum update

````
yum clean all
````

````
yum update
````

**Comment by webhost247** [comment]
@Nirjonadda Just tried doing this now the results are still the same unfortunately I keep getting told no new packages are available

![image](https://cloud.githubusercontent.com/assets/5007268/23336398/e999ef32-fbc6-11e6-8386-e95b14495b69.png)

**Comment by ubidzz** [comment]
try the codeit MAINLINE VERSION it may work because it has nginx v1.11.10 and stable use 1.10.3 and the is the same as Engintron and it seeing it don't need updating where MAINLINE VERSION would update it.
I had the same problem after unstalling my nginx and trying to reinstall it but couldn't with CodeIT and had to install Engintron first then install the codeit MAINLINE VERSION and it worked.

here is what i did

1. SSH in my server
2. Ran this code
cd /etc/yum.repos.d && wget https://repo.codeit.guru/codeit.mainline.el`rpm -q --qf "%{VERSION}" $(rpm -q --whatprovides redhat-release)`.repo
3. yum clean all
4. yum update

and it updated the nginx server using the CodeIT repo.

**Comment by tux-tn** [comment]
@webhost247 your issue is related to the fact that you have another repo providing nginx package
follow those steps (as root) to get nginx installed:
```bash
rm /etc/yum.repos.d/nginx.repo
rpm -e --justdb --nodeps nginx
yum update
yum install nginx
```
and of course check if the correct version is installed using `nginx -V`

**Comment by webhost247** [comment]
@tux-tn 
Praise the sun..

> [root@server nginx]# nginx -V
nginx version: nginx/1.11.10
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-11) (GCC)
built with OpenSSL 1.0.2k  26 Jan 2017

Could I please ask do I switch over to HTTP/2 now?

**Comment by tux-tn** [comment]
Just restart nginx if it's not already done and you are good.. 
You can check if h2 is working using network tab of your browser devtools or  by using ssllabs server test

**Comment by webhost247** [comment]
@tux-tn
Thank you for the help, sites under my server show h2 as the protocol now.

**Comment by tux-tn** [comment]
You are welcome, i am glad that it helped (up)

**Comment by Nirjonadda** [comment]
@tux-tn Why i am still using cPanel OpenSSL 1.0.1e-fips 11 Feb 2013? How to built with OpenSSL 1.0.2k 26 Jan 2017?

````
[root@na ~]# nginx -V
nginx version: nginx/1.10.3
built by gcc 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC)
built with OpenSSL 1.0.1e-fips 11 Feb 2013
TLS SNI support enabled
````

**Comment by ubidzz** [comment]
Because you didn't edit the cPanel files that are used and it not recommended to change your cPanel stuff because it can brake your cPanel or when they release a update it will change the files. Because cPanel only release what is stable from the OS provider that will work with cPanel and is why you are still using openSSL 1.0.1e. The OpenSSL 1.0.1e will only be used in like WHM/cPanel features because they are using the apache server to handle the connections and is not compiled to use the nginx server.

**Comment by Nirjonadda** [comment]
@ubidzz As i know for this post https://github.com/engintron/engintron/issues/437#issuecomment-282514324 , I have enabled the CodeIT repo OpenSSL For nginx stable.
Updated:
apr.x86_64 0:1.5.2-1.el7.codeit
Complete!

**Comment by tux-tn** [comment]
@Nirjonadda  try to follow the steps from my comment https://github.com/engintron/engintron/issues/437#issuecomment-282548060

**Comment by Nirjonadda** [comment]
@tux-tn  Do we still Need nginx install? When we have already nginx installed with engintron. It's not do nginx install 2 time in same server. Also we have used CodeIT repo. 

Do we need do re-run CodeIT repo after nginx installation?

**Comment by tux-tn** [comment]
@Nirjonadda read #368 to understand the need of custom nginx.

1/ You are using default nginx version If your pages are still served using HTTP/1.1 => That's why i am asking your to follow the steps i mentioned 
2/ nginx will be installed using CentOS package manager => One nginx version will be installed
3/ The one provided by engintron is from nginx officiel repo and doesn't contain ALPN support on Centos => That's why you need to remove /etc/yum.repos.d/nginx.repo to let your package manager install the one provided by CodeIT repo

Feel free to ask me if you need more explanation or if something wasn't clear

**Comment by Nirjonadda** [comment]
@tux-tn If we do use custom nginx then we are no need to using Engintron? OR Engintron still do work with custom nginx? Engintron why not do supporting ALPN by default? Looking forward to Engintron development teams support.

**Comment by tux-tn** [comment]
Engintron provides custom configuration for Nginx/Apache/Cpanel. It's completely independent from which nginx build you are using.
Lack of ALPN support is due to the OpenSSL version shipped by CentOS (OpenSSL 1.0.1e), it's distribution related and compiling or installing a newer global version will break things (That's why Cpanel are not using a newer version of OpenSSL)

To address this issue, you can recompile NGINX from source and use a private build with OpenSSL 1.0.2. 
But actually, folks from CodeIT are already doing this and providing a CentOS repository containing prebuild versions of nginx compiled against OpenSSL 1.0.2.
That's why i suggested you to install CodeIT nginx repo, it's the easiest way to have ALPN support
